Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#69437 Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}**

linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

Differential Revision: [D32883666](https://our.internmc.facebook.com/intern/diff/D32883666)

zou3519 committed:
Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}

linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

[ghstack-poisoned]

zou3519 commented:
cc @Lezcano and @IvanKobzarev on the linalg changes. This is what we discussed in https://github.com/pytorch/pytorch/pull/69263

cc @albanD on the __torch_dispatch__ changes

lezcano reviewed (dismissed):
Loving this PR! Thank you so much for this!

It would be great to change all the uses of the `*CheckErrors` to this new function, and it should be as straightforward as a find/replace. Feel free to do so, or just land this and we can do that in a separate follow-up PR.

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1785,7 +1785,11 @@ void concrete_dispatch_fn(
   ));
 
-  if (op.schema().returns().size() == 1) {
+  if (num_returns == 0) {
```
Good catch!

albanD reviewed (dismissed):
SGTM thanks!

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).

zou3519 commented:
> It would be great to change all the uses of the `*CheckErrors` to this new function, and it should be as straightforward as a find/replace. Feel free to do so, or just land this and we can do that in a separate follow-up PR.

I'm going to land this first to unblock functorch but I'm happy to do the follow-up on top of this

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -735,6 +735,17 @@ template<> void blasTriangularSolve<float>(char side, char uplo, char trans, cha
 }
 #endif
 
+void _linalg_check_errors(
+    const Tensor& info,
+    const c10::string_view api_name,
+    bool is_matrix) {
+  if (is_matrix) {
+    singleCheckErrors(info.item<int64_t>(), api_name);
+  } else {
+    batchCheckErrors(info, api_name);
+  }
+}
```
On second thought... does `is_matrix` need to be an argument to this function? Could we infer `is_matrix` from `info.numel() == 1`? cc @Lezcano @IvanYashchuk

Lezcano replied:
We can't do that because matrices of size `(1, 1, 3, 2)` and `(3, 2)` would be treated the same.
Now, we could probably infer from `info.dim() == 0`. What do you think @IvanYashchuk ?

IvanYashchuk replied:
I don't know if `info` is consistently implemented to have shape `()` for single input matrices, it could also be `(1,)`.

zou3519 replied:
@Lezcano @IvanYashchuk another question: is the condition always "is a matrix"? Looking at some code (I decided to change usages of batchCheckErrors to this new logic in this PR to avoid churn), there are some cases where the condition is "result.dim() > 1":

https://github.com/pytorch/pytorch/blob/9aa1b3e396c1fe9dadecaee04c9321f659d863ac/aten/src/ATen/native/BatchLinearAlgebra.cpp#L962-L976

Should I rename (and invert) `is_matrix` to `is_batch` instead?

IvanYashchuk replied:
@zou3519, thank you for pointing out this use case that deviates from other uses of `info` to check the error! It was necessary here because input could be a single matrix but then internally it was expanded to be batched which affects the size of `infos`.

I submitted a pull request (https://github.com/pytorch/pytorch/pull/69752) fixing that and now the condition is `if (input.dim() > 2)` instead of `if (vector_case ? result.dim() > 1 : result.dim() > 2)`.

> Should I rename (and invert) is_matrix to is_batch instead?

I don't think it's necessary, only if you prefer "is_batch".

Lezcano commented:
What's the state of this PR @zou3519 ?

zou3519 commented:
> What's the state of this PR @zou3519 ?

Oops, I'll clean this up and submit an update early next week

zou3519 committed:
Update on "Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}"


linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

Differential Revision: [D32883666](https://our.internmc.facebook.com/intern/diff/D32883666)

[ghstack-poisoned]

zou3519 committed:
Update on "Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}"


linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

Differential Revision: [D32883666](https://our.internmc.facebook.com/intern/diff/D32883666)

[ghstack-poisoned]

zou3519 commented:
This is ready for re-review. What changed since the last iteration is that I incorporated @Lezcano's suggestion to "change all the uses of the *CheckErrors to this new function".

NB: I wasn't able to get rid of all the batchCheckErrors callsites. There are three callsites that that use `vector<int64_t> infos` instead of Tensor infos
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1267-L1275
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L2525-L2529
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L3068-L3072

None of these matter for composite compliance though since they occur in non-composite functions.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1176,14 +1167,8 @@ Tensor& linalg_inv_out(const Tensor &input, Tensor &result) {
-  }
-
+  at::_linalg_check_errors(infos_lu, "linalg.inv", result.dim() == 2);
+  at::_linalg_check_errors(infos_getri, "linalg.inv", result.dim() == 2);
```
[Unrelated to this PR] @IvanYashchuk @nikitaved We should just check the infos of LU here, as if they are fine, getri should always succeed.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6945,6 +6945,10 @@
 
+- func: _linalg_check_errors(Tensor info, str api_name, *, bool is_matrix) -> ()
+  dispatch:
+    CompositeExplicitAutograd: _linalg_check_errors
```
Why is this needed?

zou3519 replied:
We're hiding the data_ptr calls in a new non-composite operator, `_linalg_check_errors`. Does that answer the question or were you referring to something specific in these three lines?

albanD replied:
Ho I was was refering to making this CompositeExplicitAutograd vs the default CompositeImplicitAutograd

zou3519 replied:
Making this CompositeImplicitAutograd makes it so that this operation is composite with respect to autograd, which is a problem because internally it calls .data_ptr().

Making this CompositeExplicitAutograd makes it so that this operation is not composite with respect to autograd (I think? The naming is really weird) so that the .data_ptr() call is hidden.

Does that sound right?

albanD replied:
Sounds good.

zou3519 replied:
Side note: can we rename CompositeExplicitAutograd? I'm not sure what it's supposed to mean

albanD replied:
This is the 3rd name for it already....
The idea is:
- CompositeImpliciteAutograd: is a composite op (used by all backends) that is autodiffed (has an implicit autograd).
- CompositeExplicitAutograd: is a composite op (used by all backends) that has an explicit autograd formula.

zou3519 replied:
I see, so CompositeExplicitAutograd means it's composite w.r.t. to the backend, not that it is composite w.r.t autograd :P

albanD replied:
Yes, the composite part is about aten.
implicit/explicit is about autograd.

We did discussed these names at length with Ailing. But it is very very hard to find a name that makes everyone happy :(

albanD reviewed (approved):
SGTM

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).

zou3519 committed:
Update on "Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}"


linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

Differential Revision: [D32883666](https://our.internmc.facebook.com/intern/diff/D32883666)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).
